### PR TITLE
fix(aggregate): give _ReportChangeView an entity_id field

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

`main` is currently red on the non-canonical unit-test lanes: `abicheck.workflows.aggregate.reconcile._ReportChangeView` never got a `entity_id` field, so any `aggregate` finding-matrix computation over a report whose `Change` carried a live `entity_id` raises `AttributeError`.

## Motivation

ADR-063 Phase 2 (#941-ish series) gave `Change` an `entity_id` field, and a later commit (`a589f8a7a feat(compare): resolve_change_identity consumes Change.entity_id`) taught `finding_identity.resolve_change_identity` to read `change.entity_id` unconditionally (gated only on the pre-existing `is_batch` check). `abicheck/workflows/aggregate/reconcile.py`'s `_ReportChangeView` — the read-back adapter `resolve_report_change_identity` builds from a report's own JSON and passes to that same function via `cast("Change", view)` — was never updated with a matching field. Since `resolve_change_identity` is called unconditionally on every entry, this breaks **every** `aggregate` finding-matrix computation, not a corner case: 105 tests fail on `main` today across `tests/test_aggregate.py` and `tests/test_aggregate_findings.py`.

I found this while investigating an unrelated CI failure on a different open PR whose branch predates this regression; confirmed reproducing on a clean `origin/main` checkout with no other changes.

## Changes

- `abicheck/workflows/aggregate/reconcile.py`: add `entity_id: EntityId | None` to `_ReportChangeView`, always passed as `None` in `resolve_report_change_identity` — like `qualified_name`, `_change_to_dict` never serializes `entity_id`, so a consistent loss across every report is fine for the cross-report comparison this adapter exists to serve (the field is documented as additive-only, never affecting `primary_id`/`tier`).
- `tests/test_report_change_view_entity_id.py` (new file, rather than growing the debt-tracked `tests/test_aggregate_findings.py`): proves the round trip through a *real* `Change` → `_change_to_dict` → `resolve_report_change_identity` no longer raises, and that `entity_id` doesn't change the round-tripped identity relative to an otherwise-identical finding with none.
- `tests/regressions/manifest.py`: new `BugClass` entry (`adapter.duck_typed_view_attribute_drift`) generalizing the invariant — a duck-typed read-back adapter presented as a richer type must be updated in lockstep whenever the consuming function starts reading a new attribute unconditionally, or every call through it breaks uniformly.
- Changelog fragment via `scriv create`.

## Bug-fix test contract

- Bug class: `adapter.duck_typed_view_attribute_drift` (new entry, `tests/regressions/manifest.py`)
- Publicly observable failure: `aggregate --report-dir ... --expect ...` (and the underlying `python -m abicheck.aggregate` / `resolve_report_change_identity` API) raises `AttributeError: '_ReportChangeView' object has no attribute 'entity_id'` on any finding at all — not a specific input, every finding.
- Regression test fails on base: yes — `tests/test_report_change_view_entity_id.py::test_a_live_changes_entity_id_survives_the_report_round_trip` raises the exact `AttributeError` against unpatched `main`.
- Negative control: `test_a_report_already_on_the_current_schema_is_left_alone`-style non-crashing behavior is unaffected — every pre-existing passing test in `tests/test_aggregate_findings.py`/`tests/test_aggregate.py` (307 tests) still passes unchanged.
- Public-surface test: none added here — the new test calls `resolve_report_change_identity`/`_change_to_dict` directly (internal modules), same as the rest of `tests/test_aggregate_findings.py`; `tests/test_aggregate.py`'s existing CLI-level tests already exercise `aggregate`'s real CLI surface and now pass again.
- Axes covered: n/a (single, uniform failure mode — no per-input variation to cover; the fix and test both exercise the real `Change` dataclass → `_change_to_dict` → `resolve_report_change_identity` pipeline rather than a hand-built dict literal, since a hand-built dict wouldn't prove a *live* `Change`'s `entity_id` survives the round trip).
- General invariant: `_ReportChangeView` must expose every attribute `finding_identity.resolve_change_identity` reads on a `Change`, or every call through `resolve_report_change_identity` breaks uniformly the moment that function starts reading a new one — stated as the new `adapter.duck_typed_view_attribute_drift` bug class in `tests/regressions/manifest.py` so a future PR extending `resolve_change_identity`'s attribute surface has a home to check first.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] Changelog fragment added (`scriv create`)
- [x] Docs updated if user-visible behaviour changed — n/a, internal-only fix
- [x] `pre-commit run --all-files` passes locally (ruff check/format, mypy, `check_architecture.py`, `check_ai_readiness.py` all clean)
- [x] No new `mypy` or `ruff` errors introduced

---
_Generated by [Claude Code](https://claude.ai/code/session_016T72p7RoCCiwVcnYpvPRPR)_

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed report change identity handling so changes with an entity ID round-trip correctly without altering their identity.
  - Improved detection of indirectly accessed change attributes, including aliases, helper calls, positional-only parameters, and nested scopes.

- **Tests**
  - Added regression coverage for entity ID preservation.
  - Expanded structural checks to ensure report data includes every change attribute required during identity resolution.
  - Registered a regression case for stale duck-typed adapter failures.

### CODE REVIEW

File: tests/test_report_change_view_entity_id.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
+    """A real `Change` carrying an `entity_id` must round-trip through a
+    report without raising, and `entity_id` (never serialized by
+    `_change_to_dict`) must not change the round-tripped identity relative
+    to an otherwise-identical finding that never had one."""
+    eid = EntityId(scope=(), kind=EntityKind.FUNCTION, leaf_name="add")
+    with_id = resolve_report_change_identity(
+        _change_to_dict(Change(entity_id=eid, **_KWARGS))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in two commits: 93e62491 registered the `adapter.duck_typed_view_attribute_drift` `BugClass` entry in `tests/regressions/manifest.py` (this landed slightly after the commit you reviewed, `0a5a5aa`), and 6a0b8293 adds the generalized structural test you asked for — `test_report_change_view_covers_every_attribute_resolve_change_identity_reads` parses `finding_identity.py`'s real source via `ast`, statically discovers every `change.<attr>` read reachable from `resolve_change_identity` (following calls into its same-module helpers), and asserts `_ReportChangeView`'s field set is a superset. Verified as load-bearing by swapping in the pre-fix `reconcile.py`: both the concrete and the new structural test fail, the latter naming the exact missing attribute.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    if func is None:
+        return set()
+    attrs: set[str] = set()
+    for node in ast.walk(func):
+        if (
+            isinstance(node, ast.Attribute)
+            and isinstance(node.value, ast.Name)
+            and node.value.id == param_name
+        ):
+            attrs.add(node.attr)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in e85bf827: `_change_attrs_read` now also recognizes `getattr(<param>, "name", ...)` where the attribute name is a string literal, alongside the original `change.attr` syntax. Added `test_change_attrs_read_catches_an_indirect_getattr_field_read`, a dedicated fixture proving it — verified as a real gap by running the pre-fix walker against the identical fixture and confirming it returned an empty set (no `future_field` found) before the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    func = funcs_by_name.get(func_name)
+    if func is None:
+        return set()
+    attrs: set[str] = set()
+    for node in ast.walk(func):
+        if (
+            isinstance(node, ast.Attribute)
+            and isinstance(node.value, ast.Name)
+            and node.value.id == param_name
+        ):
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in ad6354161: `_change_attrs_read` now tracks simple direct-name-assignment aliases of the tracked parameter within the same function body (`candidate = change`), following an alias of an alias transitively, and matches both a direct attribute read off the alias and a call-argument forward through any tracked name (not just the original parameter). Added `test_change_attrs_read_catches_a_direct_local_alias_read` and `test_change_attrs_read_catches_a_local_alias_forwarded_to_a_helper` — verified as real gaps by running the pre-fix walker against both fixtures and confirming it returned an empty set for each before the fix.

Note: this PR's branch also picked up a merge conflict against `main` after PR #956 merged (its own branch had independently ported a simpler version of this same `entity_id` fix) — resolved by keeping this branch's fuller structural-guard version, which is a strict superset of main's copy.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 4]
```diff
@@ -56,9 +73,200 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    visited.add(key)
+    func = funcs_by_name.get(func_name)
+    if func is None:
+        return set()
+    tracked = {param_name}
+    for node in ast.walk(func):
+        if (
+            isinstance(node, ast.Assign)
+            and isinstance(node.value, ast.Name)
+            and node.value.id in tracked
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in 2253d5453: the alias-tracking pass now also matches a name-targeted `ast.AnnAssign` whose value is an already-tracked name (`candidate: Change = change`), alongside the existing `ast.Assign` case. Added `test_change_attrs_read_catches_an_annotated_local_alias_read` — verified as a real gap by running the pre-fix walker against the fixture and confirming it returned an empty set before the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 5]
```diff
@@ -56,9 +73,230 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+            and isinstance(node.args[1].value, str)
+        ):
+            attrs.add(node.args[1].value)
+        elif (
+            isinstance(node, ast.Call)
+            and isinstance(node.func, ast.Name)
+            and node.func.id in funcs_by_name
+        ):
+            callee = funcs_by_name[node.func.id]
+            arg_names = [a.arg for a in callee.args.args]
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in a27e77198: the helper-call binding now reads `callee.args.posonlyargs` followed by `callee.args.args` (the same order Python itself uses for positional binding), not just `args.args`. Added `test_change_attrs_read_follows_a_positional_only_helper_parameter` — verified as a real gap by running the pre-fix walker against the fixture and confirming it returned an empty set before the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 6]
```diff
@@ -56,9 +73,258 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+                    tracked.add(target.id)
+        elif (
+            isinstance(node, ast.AnnAssign)
+            and isinstance(node.target, ast.Name)
+            and isinstance(node.value, ast.Name)
+            and node.value.id in tracked
+        ):
+            tracked.add(node.target.id)
+    attrs: set[str] = set()
+    for node in ast.walk(func):
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in 3352eb861: added `_walk_own_scope`, a stack-based walk that behaves like `ast.walk` for the node passed in but never descends past a nested `FunctionDef`/`AsyncFunctionDef`/`Lambda`/`ClassDef` boundary, and used it in place of both `ast.walk(func)` calls in `_change_attrs_read`. Added `test_change_attrs_read_ignores_a_nested_functions_shadowed_parameter` — confirmed as a real gap by running the pre-fix walker against the fixture and observing the misattributed read, and spot-checked that a genuine outer-scope read is still found unchanged after the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 7]
```diff
@@ -56,9 +74,309 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    scope are excluded, along with everything inside them.
+    """
+    stack = [node]
+    while stack:
+        current = stack.pop()
+        yield current
+        for child in ast.iter_child_nodes(current):
+            if child is not node and isinstance(child, _SCOPE_BOUNDARY_NODES):
+                continue
+            stack.append(child)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:304]: Fixed in ac80e80fd: alias tracking now runs to a fixed point (loop until no new name is added) instead of a single pass, so the result no longer depends on traversal order at all. Added `test_change_attrs_read_follows_a_multi_hop_alias_chain` — verified as a real gap by running the pre-fix walker against the fixture and observing an empty set before the fix.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 8]
```diff
@@ -56,9 +74,309 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    of its own kind -- only nodes strictly *inside* it that open a new
+    scope are excluded, along with everything inside them.
+    """
+    stack = [node]
+    while stack:
+        current = stack.pop()
+        yield current
+        for child in ast.iter_child_nodes(current):
+            if child is not node and isinstance(child, _SCOPE_BOUNDARY_NODES):
+                continue
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in ac80e80fd: replaced the blanket scope-skip with `_iter_scope_aware`, which threads a per-node set of names actually *shadowed* by an enclosing nested function/lambda's own parameter list (posonly/regular/kwonly/`*args`/`**kwargs`) — only a name a nested scope truly rebinds is excluded there; every other name, including the tracked one via genuine closure, stays visible. Added `test_change_attrs_read_counts_a_genuine_closure_capture` — verified as a real gap by running the pre-fix walker against the fixture and observing an empty set — and spot-checked that the sixth round's own shadowing fixture still correctly resolves to an empty set after this change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 9]
```diff
@@ -56,9 +74,388 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    no parameter list to rebind a name via) -- a name reassigned inside
+    it is a rarer, unmodeled edge matching this helper's existing "small,
+    targeted, not general-purpose" scope.
+    """
+    stack: list[tuple[ast.AST, frozenset[str]]] = [(node, frozenset())]
+    while stack:
+        current, shadowed = stack.pop()
+        yield current, shadowed
+        if current is not node and isinstance(current, _PARAM_SCOPE_NODES):
+            shadowed = shadowed | _own_param_names(current.args)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in 57d54624c: `_iter_scope_aware` now walks a nested function/lambda's `args` (and, for a real function, `decorator_list`/`returns`) with the *enclosing* scope's shadow set, and only `body` with the new, shadowed one — matching Python's actual scoping rules instead of applying one shadow set to the whole subtree. Added `test_change_attrs_read_evaluates_a_nested_default_in_the_enclosing_scope` — verified as a real gap by running the pre-fix walker against the fixture and observing an empty set, and spot-checked every earlier fixture (alias chains, closure capture, shadowing, positional-only, indirect getattr) still resolves correctly after this change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 10]
```diff
@@ -56,9 +74,434 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    helper's existing "small, targeted, not general-purpose" scope.
+    """
+    stack: list[tuple[ast.AST, frozenset[str]]] = [(node, frozenset())]
+    while stack:
+        current, shadowed = stack.pop()
+        yield current, shadowed
+        if current is not node and isinstance(
+            current, (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda)
+        ):
+            own_shadowed = shadowed | _own_param_names(current.args)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in e35ddf0fe: added `_locally_bound_names`, which computes every name Python's scoping rules make local to a nested function/lambda by ordinary body-level binding — assignment (incl. tuple/list/starred unpacking), augmented/annotated assignment, `for`/`async for` targets, `with`/`async with ... as` targets, `except ... as` names, `import`/`from ... import ... as` aliases, walrus targets, and `global` declarations — without descending into a further-nested scope. Folded into `_iter_scope_aware`'s `own_shadowed` alongside `_own_param_names`. Deliberately excludes `nonlocal`-declared names from the shadowed set, since that statement explicitly means the name resolves to the enclosing scope rather than binding locally — pinned by its own dedicated test so a future fix doesn't accidentally treat every rebound name as shadowed.

Added `test_change_attrs_read_ignores_a_nested_functions_own_local_assignment` (confirmed as a real gap via the pre-fix walker) and `test_change_attrs_read_treats_nonlocal_as_a_genuine_capture`. Spot-checked for-loop/except/with/import/global binding forms and every earlier fixture still resolve correctly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 11]
```diff
@@ -56,9 +74,565 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    while stack:
+        current = stack.pop()
+        if not is_root and isinstance(
+            current, (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda, ast.ClassDef)
+        ):
+            continue
+        is_root = False
+        if isinstance(current, ast.Assign):
+            for assign_target in current.targets:
+                _bind_target(assign_target, names)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in 1f5b86c80: `Nonlocal` declarations are now collected into their own set and subtracted from the local-binding set only at the very end, so a sibling `Assign`/`AugAssign`/`AnnAssign` to that same name (the normal reason to declare it `nonlocal`) can never re-add it — declaration order relative to the assignment no longer matters. Added `test_change_attrs_read_keeps_a_nonlocal_capture_after_reassignment` — verified as a real gap via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 12]
```diff
@@ -56,9 +74,565 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    """
+    names: set[str] = set()
+    stack: list[ast.AST] = [scope_node]
+    is_root = True
+    while stack:
+        current = stack.pop()
+        if not is_root and isinstance(
+            current, (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda, ast.ClassDef)
+        ):
+            continue
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:261]: Fixed in 1f5b86c80: a nested `FunctionDef`/`ClassDef` now has its own `.name` recorded as a local binding in the containing scope before `continue`-ing past its separate body (Lambda has no name, handled separately, still just `continue`s). Added `test_change_attrs_read_treats_a_nested_def_name_as_a_local_binding` — verified as a real gap via the pre-fix walker, and spot-checked a nested `ClassDef` binding the same way too.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:261]: Same finding Codex raised on this commit — fixed in 1f5b86c80 the same way your suggestion describes (recording `FunctionDef`/`AsyncFunctionDef`/`ClassDef.name` before continuing past the nested scope), plus a regression test (`test_change_attrs_read_treats_a_nested_def_name_as_a_local_binding`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 13]
```diff
@@ -56,9 +74,645 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+            and node.func.id in funcs_by_name
+        ):
+            callee = funcs_by_name[node.func.id]
+            arg_names = [a.arg for a in (*callee.args.posonlyargs, *callee.args.args)]
+            for i, arg in enumerate(node.args):
+                if (
+                    isinstance(arg, ast.Name)
+                    and arg.id in tracked
+                    and arg.id not in shadowed
+                    and i < len(arg_names)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in 8bc474f66: added `_unwrap_transparent` (recognizes both bare `cast(...)` and qualified `typing.cast(...)` call forms) and applied it at every site that matches an `ast.Name` against `tracked`/`shadowed` — call arguments (positional and keyword), alias `Assign`/`AnnAssign` values, direct `Attribute` access, and `getattr`'s first argument — closing the whole class in one pass. Added `test_change_attrs_read_unwraps_a_cast_wrapped_helper_argument` — verified as a real gap via the pre-fix walker, and spot-checked the other three call sites (direct attribute, alias assignment, getattr) too.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 14]
```diff
@@ -56,9 +74,645 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+            else:
+                for decorator in current.decorator_list:
+                    stack.append((decorator, shadowed))
+                if current.returns is not None:
+                    stack.append((current.returns, shadowed))
+                for stmt in current.body:
+                    stack.append((stmt, own_shadowed))
+            continue
+        for child in ast.iter_child_nodes(current):
+            stack.append((child, shadowed))
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:392]: Fixed in 8bc474f66: `_iter_scope_aware` now has an explicit branch for `ListComp`/`SetComp`/`GeneratorExp`/`DictComp` — collects every generator's target into a comprehension-local shadow set, walks the *first* generator's own iterable expression with the enclosing shadow set (Python evaluates it there, mirroring the eighth round's default-value treatment) and everything else (later iterables, `if`s, `elt`/`key`/`value`) with the comprehension-local one. Added `test_change_attrs_read_ignores_a_comprehension_local_target` — verified as a real gap via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 15]
```diff
@@ -56,9 +74,739 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+                    tracked.add(node.target.id)
+                    changed = True
+    attrs: set[str] = set()
+    for node, shadowed in _iter_scope_aware(func):
+        if isinstance(node, ast.Attribute):
+            attr_value = _unwrap_transparent(node.value)
+            if (
+                isinstance(attr_value, ast.Name)
+                and attr_value.id in tracked
+                and attr_value.id not in shadowed
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:534]: Fixed in fb816ad3c: `_locally_bound_names` no longer collects a plain `Assign`/`AnnAssign` target whose value (after unwrapping a transparent `cast(...)`) is a bare `Name` — an alias-shaped assignment now defers entirely to the fixed-point alias-tracking pass instead of being pre-emptively shadowed. This accepts a narrower, deliberately safer trade-off (documented in the docstring): a genuinely unrelated rebind through a bare name is also no longer shadowed, over-reporting rather than silently dropping a read — favoring the false-positive direction, which this module has consistently preferred over a missed real read. `AugAssign` stays excluded from the exemption unconditionally, preserving the ninth round's own test unchanged.

Added `test_change_attrs_read_keeps_a_closure_local_alias_of_the_tracked_object` (confirmed as a real gap via the pre-fix walker) and `test_change_attrs_read_still_shadows_an_unrelated_local_via_augassign`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 16]
```diff
@@ -56,9 +74,809 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    func = funcs_by_name.get(func_name)
+    if func is None:
+        return set()
+    tracked = {param_name}
+    changed = True
+    while changed:
+        changed = False
+        for node, shadowed in _iter_scope_aware(func):
+            if isinstance(node, ast.Assign):
+                assign_value = _unwrap_transparent(node.value)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:480]: Fixed in 24077cc49: extended `_unwrap_transparent` to also unwrap a `NamedExpr` to its value (looping, so it composes with `cast(...)` in either order), and added a `NamedExpr` branch to the alias-tracking fixed-point loop mirroring the existing `Assign`/`AnnAssign` branches. Also applied the twelfth round's alias exemption to `_locally_bound_names`'s own pre-existing `NamedExpr` handling, which was unconditionally shadowing a walrus target the same way plain `Assign` once did.

Added `test_change_attrs_read_unwraps_a_walrus_wrapped_direct_attribute` and `test_change_attrs_read_tracks_a_walrus_alias_for_a_later_separate_read` (covering both top-level and nested-closure placements) — both confirmed as real gaps via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 17]
```diff
@@ -56,9 +74,892 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+                    and ann_value.id not in shadowed
+                    and node.target.id not in tracked
+                ):
+                    tracked.add(node.target.id)
+                    changed = True
+            elif isinstance(node, ast.NamedExpr) and isinstance(node.target, ast.Name):
+                walrus_value = _unwrap_transparent(node.value)
+                if (
+                    isinstance(walrus_value, ast.Name)
+                    and walrus_value.id in tracked
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:504]: Fixed in 3925c8590: added a `Match` branch to the alias-tracking fixed-point loop — when a match statement's subject (after unwrapping `cast(...)`/walrus) is a tracked, unshadowed `Name`, each `case` whose pattern is a top-level `MatchAs` with no nested pattern (a bare `case NAME:` capture, which always matches) has its name added to `tracked`. Only the simple top-level shape is recognized, not a capture nested inside a class/sequence/mapping pattern — documented as such, matching this helper's existing "small, targeted" scope. Added `test_change_attrs_read_tracks_a_match_case_capture_alias` — verified as a real gap via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 18]
```diff
@@ -56,9 +74,939 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+    while stack:
+        current, shadowed = stack.pop()
+        yield current, shadowed
+        if current is not node and isinstance(
+            current, (ast.FunctionDef, ast.AsyncFunctionDef, ast.Lambda)
+        ):
+            own_shadowed = (
+                shadowed
+                | _own_param_names(current.args)
+                | _locally_bound_names(current)
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py:361]: Fixed in e4fbe2f45: `_locally_bound_names` now tracks alias-shaped bindings (value is a bare `Name` after unwrapping) in a separate set and subtracts it from the shadow set at the very end, rather than deciding per-assignment. This module has no per-statement control-flow ordering to know whether a read happened before or after a later rebind, so a name that was EVER a genuine alias anywhere in the scope now never risks hiding a real read there again — widening the twelfth round's exemption to the whole scope instead of the single assignment. Added `test_change_attrs_read_reads_an_alias_before_a_later_unrelated_rebind` — verified as a real gap via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_report_change_view_entity_id.py [hunk 19]
```diff
@@ -56,9 +74,939 @@ def test_a_live_changes_entity_id_survives_the_report_round_trip() -> None:
...
+):
+    """Structural enforcement of this module's own stated bug class: parse
+    `finding_identity.py`'s real source (not a hand-copied attribute list
+    that could itself go stale) and fail if `resolve_change_identity` --
+    transitively, through its own same-module helpers -- reads any `Change`
+    attribute `_ReportChangeView` does not declare."""
+    tree = ast.parse(inspect.getsource(finding_identity))
+    funcs_by_name = {
+        node.name: node for node in ast.walk(tree) if isinstance(node, ast.FunctionDef)
+    }
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_report_change_view_entity_id.py]: Fixed in e4fbe2f45: added `_group_funcs_by_name`, building `dict[str, list[FunctionDef]]` instead of the flat `dict[str, FunctionDef]` used since round one, and split `_change_attrs_read` into an outer multi-candidate dispatcher (unions every same-named candidate's results, `visited` now keyed on `id(function)` rather than name) plus `_change_attrs_read_in_function` (the existing single-function analysis, otherwise unchanged). All ~20 call sites across this file's fixtures now build the grouped map via the shared helper. Added `test_change_attrs_read_follows_every_same_named_helper_candidate` — verified as a real gap via the pre-fix walker.

---
_Generated by [Claude Code](https://claude.ai/code)_
